src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs

```diff
@@ -0,0 +1,189 @@
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.Azure.WebJobs.Script.Configuration;
+using Microsoft.Azure.WebJobs.Script.Properties;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using NuGet.Versioning;
+
+namespace Microsoft.Azure.WebJobs.Script.ExtensionBundle
+{
+    public class ExtensionBundleManager : IExtensionBundleManager
+    {
+        private readonly IEnvironment _environment;
+        private readonly ExtensionBundleOptions _options;
+        private readonly ILogger _logger;
+        private readonly string _cdnUri;
+
+        public ExtensionBundleManager(ExtensionBundleOptions options, IEnvironment environment, ILoggerFactory loggerFactory)
+        {
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _logger = loggerFactory.CreateLogger(ScriptConstants.LogCategoryExtensionBundles) ?? throw new ArgumentNullException(nameof(loggerFactory));
+            _cdnUri = _environment.GetEnvironmentVariable(EnvironmentSettingNames.AlternateCdnUri) ?? ScriptConstants.CdnBaseUri;
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public bool IsExtensionBundleConfigured()
+        {
+            return !string.IsNullOrEmpty(_options.Id) && !string.IsNullOrEmpty(_options.Version?.OriginalString);
+        }
+
+        public async Task<string> GetExtensionBundle(HttpClient httpClient = null)
+        {
+            if (httpClient == null)
+            {
+                using (httpClient = new HttpClient())
+                {
+                    return await GetBundle(httpClient);
+                }
+            }
+            else
+            {
+                return await GetBundle(httpClient);
+            }
+        }
+
+        private async Task<string> GetBundle(HttpClient httpClient)
+        {
+            string latestBundleVersion = await GetLatestMatchingBundleVersion(httpClient);
+            if (string.IsNullOrEmpty(latestBundleVersion))
+            {
+                return null;
+            }
+
+            bool bundleFound = TryLocateExtensionBundle(out string bundlePath);
+            string bundleVersion = Path.GetFileName(bundlePath);
+
+            if (_environment.IsPersistentFileSystemAvailable()
+                && (!bundleFound || (Version.Parse(bundleVersion) < Version.Parse(latestBundleVersion) && _options.EnsureLatest)))
+            {
+                bundlePath = await DownloadExtensionBundleAsync(latestBundleVersion, httpClient);
+            }
+            return bundlePath;
+        }
+
+        internal bool TryLocateExtensionBundle(out string bundlePath)
+        {
+            bundlePath = null;
+            string bundleMetatdataFile = null;
+            var paths = new List<string>(_options.ProbingPaths)
```

Comment:
move this to constructor

Reply:
It is only used for this method and is a temporary construct. Do we want to add as a instance variable and alive for the entire lifetime of the object?